### PR TITLE
Add Continuous Integration using Github Actions

### PR DESCRIPTION
**Pull request by cuminandpaprika**
Addresses https://github.com/anz-bank/sysl-go/issues/6

This PR implements some basic CI on the sysl-go repo
- Linting
   - It uses golangci-lint 1.23.8 for linting via the [Reviewdog Github Action](https://github.com/reviewdog/action-golangci-lint) as golangci.com is shutting down on the 15th April [blog post](https://medium.com/golangci/golangci-com-is-closing-d1fc1bd30e0e)
- Test Coverage
    - Runs `make check-coverage` to ensure test coverage is sufficient

Other steps explored in https://github.com/anz-bank/sysl-go/issues/6 are not addressed by this PR and will be added in subsequent PRs

### CODE REVIEW

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        uses: actions/setup-go@v1
+        with:
+          go-version: 1.13
+        id: go
+
+      - name: Check out code into the Go module directory
+        uses: actions/checkout@v2
+
+      - name: Get project dependencies
+        run: go get -v -t -d ./...
```
Comment (juliaogris) [inline .github/workflows/go.yml]: Same: `go get` is a pre go modules directive, you shouldn't have to run this.

File: .github/workflows/go.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+      - name: Run tests
+        run: go test -cover -tags codeanalysis -coverprofile=coverage.out ./...
+
+      - name: Publish test results to Codecov
```
Comment (juliaogris) [inline .github/workflows/go.yml]: I think there is no need to publish this to Codecov. Do it with the makefile.

File: codecov.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+coverage:
+  precision: 2
+  round: down
+  range: "50...100"
+ignore:
+  - codegen/tests/*
```
Comment (juliaogris) [inline codecov.yml]: EOL on EOF.

Comment (juliaogris) [reply codecov.yml]: Here is why: https://gist.github.com/camh-/1bebfcff1b0f814e9b191edc60d5206b

**Review by juliaogris — COMMENTED** [review]
Ok, I've looked at this.
I'd say, avoid unnecessary external systems, especially codecov/coverall and linting (not sure about sonarcube).
I'd also say use the Makefile in the github actions as well so you can simulate a CI run locally with make.

File: .github/workflows/go.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Go 1.13
+        uses: actions/setup-go@v1
+        with:
+          go-version: 1.13
+        id: go
+
+      - name: Check out code into the Go module directory
+        uses: actions/checkout@v2
```
Comment (juliaogris) [inline .github/workflows/go.yml:34]: Is it really necessary to checkout the code on every action "job".

**Comment by cuminandpaprika** [comment]
> Ok, I've looked at this.
> I'd say, avoid unnecessary external systems, especially codecov/coverall and linting (not sure about sonarcube).
That's a good point, I've already run into some issues with some of the external services. I'll look into printing the codecoverage locally, thanks!

> I'd also say use the Makefile in the github actions as well so you can simulate a CI run locally with make.
Sounds good, definitely agree on the use of the Makefile to enable you to run as much as possible locally.

File: .golangci.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  deadline: 2m
+  build-tags:
+    - codeanalysis
+
+# golangci.com configuration
+# https://github.com/golangci/golangci/wiki/Configuration
+service:
+  golangci-lint-version: 1.19.x # use the fixed version to not introduce new linters unexpectedly
+  prepare:
+    - echo "here I can run custom commands, but no preparation needed for this repo"
```
Comment (juliaogris) [inline .golangci.yml]: Remove the line above.

File: .golangci.yml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        - scopelint
+    - path: "config/config_test.go"
+      linters:
+        - scopelint
+    # Ignore things like "FormId should be FormID"
+    - path: codegen
+      text: "ST1003:"
+      linters:
+        - stylecheck
+    # We like underscores
```
Comment (juliaogris) [inline .golangci.yml]: Really?

Comment (cuminandpaprika) [reply .golangci.yml]: haha I just copied this over from the existing .golangci.yml file in `server-lib`, so can't comment on the accuracy of liking underscores :)

File: codecov.yml (hunk 3 above)
Comment (cuminandpaprika) [reply codecov.yml]: Thanks, removed the codecov file

File: .github/workflows/go.yml (hunk 4 above)
Comment (cuminandpaprika) [reply .github/workflows/go.yml:34]: I believe so, jobs run in parallel in separate environments so each one will need to checkout to code independently

Edit: Clearer explanation here https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow#using-the-checkout-action

Docs here
https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobs

File: .golangci.yml (hunk 5 above)
Comment (cuminandpaprika) [reply .golangci.yml]: Removed the entire service section, thanks!

File: .github/workflows/go.yml (hunk 2 above)
Comment (cuminandpaprika) [reply .github/workflows/go.yml]: Yup, agreed, I've removed the dependency on Codecov (I think it just adds noise to the PR)

File: codecov.yml (hunk 3 above)
Comment (cuminandpaprika) [reply codecov.yml]: Fixed, thanks!

**Comment by cuminandpaprika** [comment]
@juliaogris 

Here's what I've gone with instead:
- Update the makefile so that CI can run make commands
- Uses [this](https://github.com/actions-contrib/golangci-lint ) action to run golangci-lint
    - I didn't see a need to make our own, and this one allows the version of golangci-lint to be specified
    - Potentially, something like https://github.com/marketplace/actions/golangci-lint-action would be great, as it posts annotations back to github, which link to the line of code that needs to be fixed. But it doesn't support specifying a version of golangci-lint
- Remove codecov and run `make test-coverage` as suggested

- [x] Still need to fix the linting issues

File: .github/workflows/go.yml (hunk 4 above)
Comment (juliaogris) [reply .github/workflows/go.yml:34]: Thanks.

File: .golangci.yml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+linters:
+  enable-all: true
+  disable:
+    - dupl
+    - gochecknoglobals
+    - gocyclo
+    - goconst
+    - lll
+    - misspell
```
Comment (juliaogris) [inline .golangci.yml]: that's a good one to have imo.

In fact I'd keep:
lll, misspell, gochecknoglobals, and gocyclo.
You can update gocyclo to higher values in gocyclo specific settings;
you can also "exclude" individual lines / files with //nolint:lll for instance

File: Makefile [hunk 8]
```diff
@@ -1,5 +1,52 @@
...
+.PHONY: all clean
+
+# -- Lint ----------------------------------------------------------------------
+lint: ## Lint Go Source Code
+	golangci-lint run
+
+tidy:
+	go mod tidy
+
+.PHONY: lint tidy clean
```
Comment (juliaogris) [inline Makefile]: no clean target here?

File: validator/validator_test.go [hunk 9]
```diff
@@ -135,7 +135,7 @@ func testDummyValidator(reflect.Value) interface{} {
 	return 6
 }
 
-//nolint:gochecknoinits
+//nolint:gochecknoinits // We ignore these for tests
 func init() {
```
Comment (juliaogris) [inline validator/validator_test.go]: Maybe this should / could be done in [TestMain](https://golang.org/pkg/testing/#hdr-Main)?

File: README.md [hunk 10]
```diff
@@ -14,19 +14,30 @@ Go get the repository
 
 Ensure your environment provides:
 
-- [go 1.12.9](https://golang.org/)
-- [golangci-lint 1.17.1](https://github.com/golangci/golangci-lint)
-- some working method of obtaining dependencies listed in `go.mod` (working internet access, `GOPROXY`)
-- env var `GOFLAGS="-mod=vendor"`
-- correctly configured cntlm or alpaca proxy running on localhost at port 3128 (only for updating vendor dependencies)
+- [go 1.13](https://golang.org/doc/install)
```
Comment (juliaogris) [inline README.md:17]: maybe time to upgrade to 1.14.1 ?

File: Makefile [hunk 11]
```diff
@@ -1,5 +1,52 @@
...
+coverage: test  ## Show test coverage in your browser
+	go tool cover -html=$(COVERFILE)
+
+clean: 
+	rm -f $(COVERFILE)
+
+CHECK_COVERAGE = awk -F '[ \t%]+' '/^total:/ && $$3 < $(COVERAGE) {exit 1}'
+FAIL_COVERAGE = { echo '$(COLOUR_RED)FAIL - Coverage below $(COVERAGE)%$(COLOUR_NORMAL)'; exit 1; }
 
+.PHONY: check-coverage coverage test
```
Comment (juliaogris) [inline Makefile:35]: `test` target missing.

File: Makefile [hunk 12]
```diff
@@ -1,5 +1,52 @@
...
+
+# --- Utilities ---------------------------------------------------------------
+COLOUR_NORMAL = $(shell tput sgr0 2>/dev/null)
+COLOUR_RED    = $(shell tput setaf 1 2>/dev/null)
+COLOUR_GREEN  = $(shell tput setaf 2 2>/dev/null)
+COLOUR_WHITE  = $(shell tput setaf 7 2>/dev/null)
+BOLD          = $(shell tput bold 2>/dev/null)
+
+help:
+	@awk -F ':.*## ' 'NF == 2 && $$1 ~ /^[A-Za-z0-9_-]+$$/ { printf "$(BOLD)$(COLOUR_WHITE)%-20s$(COLOUR_NORMAL)%s\n", $$1, $$2}' $(MAKEFILE_LIST) | sort
```
Comment (juliaogris) [inline Makefile:45]: I think it's great that you are re-using this `help` target.

Maybe also add help messages for all targets with `##`.
You can look at the output with `make help`

**Review by juliaogris — COMMENTED** [review]
I think this is quite alright as is.
However, if you do agree with some comments and you don't want to update them as part of this PR, please write an issue for later.

It would also be great to have the git history cleaned up a bit.

File: .golangci.yml (hunk 6 above)
Comment (cuminandpaprika) [reply .golangci.yml]: Removed this in the latest commits

File: Makefile (hunk 12 above)
Comment (cuminandpaprika) [reply Makefile:45]: Thanks for that, added help messages for all targets, it's a really nice feature to have in a makefile!

File: README.md (hunk 10 above)
Comment (cuminandpaprika) [reply README.md:17]: I think I'd like to keep the version of this at least consistent with Sysl, given that it's the same team of developers. Will create an issue for it though https://github.com/anz-bank/sysl-go/issues/34

File: Makefile (hunk 8 above)
Comment (cuminandpaprika) [reply Makefile]: Thanks, fixed in 728d9a010242a1d6784cb6665c4f590129a3b23a

File: .golangci.yml (hunk 7 above)
Comment (cuminandpaprika) [reply .golangci.yml]: That's a great point, 
I've updated to turn on `misspell`, `gocyclo` and `goconst`

It looks like it'd be quite a bit more work to enable `lll` and `gochecknoglobals` so I'll raise those as issues 
https://github.com/anz-bank/sysl-go/issues/33
https://github.com/anz-bank/sysl-go/issues/32

File: Makefile (hunk 11 above)
Comment (cuminandpaprika) [reply Makefile:35]: Sorry, a little bit confused by this. The target for `test` is above I think
```
test: ## Run all tests
	go test -coverprofile=$(COVERFILE) -tags codeanalysis ./...
```

File: validator/validator_test.go (hunk 9 above)
Comment (cuminandpaprika) [reply validator/validator_test.go]: Good point, fixed in e97548cb87552a5b17071bd5097d429d82f62fe5

File: Makefile (hunk 11 above)
Comment (juliaogris) [reply Makefile:35]: yeah, so `.PHONY` needs to hold _all_ targets that don't match any actual file output generation. (this is an artefact from creating/ compiling c executables). Maybe we should talk about this at some point in more detail.

Anyway, long story short, you need to list `test` under `.PHONY` too.

**Review by juliaogris — COMMENTED** [review]
This looks pretty good to me now, just needs that `.PHONY` update in the `Makefile`.
Also, would be great if you could squash your fixup commits and maybe just have three commits:
- linter updates
- makefile extension & docs update
- docs version updates
- CI
Or so....

File: Makefile (hunk 11 above)
Comment (cuminandpaprika) [reply Makefile:35]: Awesome, thanks for the explanation, `test` has been included undernder `.PHONY` in https://github.com/anz-bank/sysl-go/pull/23/commits/a459be5ee642cd9884c84430384c4e82f9e0abdd

**Comment by cuminandpaprika** [comment]
> This looks pretty good to me now, just needs that `.PHONY` update in the `Makefile`.
> Also, would be great if you could squash your fixup commits and maybe just have three commits:
> 
> * linter updates
> * makefile extension & docs update
> * docs version updates
> * CI
>   Or so....

Ah, thanks for the example, I wasn't sure what granularity is desirable for commits, that helps a lot to guide me!

Rebased and squashed, and addressed your comment on the makefile target 👍

**Review by juliaogris — APPROVED** [review]

**Pull request merged**